geth/utils/validation.py

```diff
@@ -82,18 +87,84 @@ class GethKwargs(BaseModel):
 
     model_config = ConfigDict(extra="forbid")
 
-    def set_field_if_none(self, field_name: str, value: Any) -> None:
-        if getattr(self, field_name, None) is None:
-            setattr(self, field_name, value)
-
 
 def validate_geth_kwargs(geth_kwargs: dict[str, Any]) -> bool:
     """
     Converts geth_kwargs to GethKwargs and raises a ValueError if the conversion fails.
     """
     try:
         GethKwargs(**geth_kwargs)
-    except TypeError:
-        # TODO more specific error message
-        raise ValueError("Invalid geth_kwargs")
+    except ValidationError as e:
+        raise ValueError(f"geth_kwargs validation failed: {e}")
+    except TypeError as e:
+        raise ValueError(f"error while validating geth_kwargs: {e}")
+    return True
+
+
+class GenesisDataConfig(BaseModel):
+    ethash: dict[str, Any] = {}
+    homesteadBlock: int = 0
+    daoForkBlock: int = 0
+    daoForkSupport: bool = True
+    eip150Block: int = 0
+    eip155Block: int = 0
+    eip158Block: int = 0
+    byzantiumBlock: int = 0
+    constantinopleBlock: int = 0
+    petersburgBlock: int = 0
+    istanbulBlock: int = 0
+    berlinBlock: int = 0
+    londonBlock: int = 0
+    arrowGlacierBlock: int = 0
+    grayGlacierBlock: int = 0
+    # merge
+    terminalTotalDifficulty: int = 0
+    terminalTotalDifficultyPassed: bool = True
+    # post-merge, timestamp is used for network transitions
+    shanghaiTime: int = 0
+    cancunTime: int = 0
+
+    model_config = ConfigDict(extra="forbid")
+
+
+class GenesisData(BaseModel):
+    alloc: dict[str, dict[str, Any]] = {}
+    coinbase: str = "0x3333333333333333333333333333333333333333"
+    config: dict[str, Any] = GenesisDataConfig().model_dump()
+    difficulty: str = "0x0"
+    extraData: str = (
+        "0x0000000000000000000000000000000000000000000000000000000000000000"
+    )
+    gasLimit: str = "0x47e7c4"
+    mixhash: str = "0x0000000000000000000000000000000000000000000000000000000000000000"
+    nonce: str = "0x0"
+    parentHash: str = (
+        "0x0000000000000000000000000000000000000000000000000000000000000000"
+    )
+    timestamp: str = "0x0"
+
+    model_config = ConfigDict(extra="forbid")
+
+
+def validate_genesis_data(genesis_data: GenesisDataTypedDict) -> bool:
+    """
+    Validates the genesis data config field
+    """
+    genesis_data_config = genesis_data.get("config", None)
+    if genesis_data_config:
+        try:
+            GenesisDataConfig(**genesis_data_config)
+        except ValidationError as e:
+            raise ValueError(f"genesis_data config field validation failed: {e}")
+        except TypeError as e:
+            raise ValueError(f"error while validating genesis_data config field: {e}")
+    """
+    Validates the genesis data
+    """
+    try:
+        GenesisData(**genesis_data)
+    except ValidationError as e:
+        raise ValueError(f"genesis_data validation failed: {e}")
```

Comment:
Sorry for the nit, is there a reason we want to take a `ValidationError` and raise a more generic `ValueError` from it? Do we need to catch it here instead of letting it raise? Can we re-raise as `ValidationError` with the clearer message? I'm not sure it matters too much, nor am I sure what is most ideal, just want to get more clarity on the reasoning for the except blocks here.

Reply:
I'm seagulling a little bit here (sorry @pacrob), but while we're on the error topic - does it make sense to raise a py-geth specific error here since we're moving in that direction in other libraries?

Reply:
Agreed on pygeth-specific specific error.